banyand/tsdb/shard.go

```diff
@@ -52,66 +54,234 @@ func (s *shard) Index() IndexDatabase {
+		return time.Parse(segMonthFormat, value)
+	case YEAR:
+		return time.Parse(segYearFormat, value)
+	case MILLISECOND:
+		return time.Parse(segMillisecondFormat, value)
+	}
+	panic("invalid interval unit")
+}
+
+type IntervalRule struct {
+	Unit IntervalUnit
+	Num  int
+}
+
+func (ir IntervalRule) NextTime(current time.Time) time.Time {
+	switch ir.Unit {
+	case DAY:
+		return current.AddDate(0, 0, ir.Num)
+	case MONTH:
+		return current.AddDate(0, ir.Num, 0)
+	case YEAR:
+		return current.AddDate(ir.Num, 0, 0)
+	case MILLISECOND:
+		return current.Add(time.Millisecond * time.Duration(ir.Num))
+	}
+	panic("invalid interval unit")
+}
+
+type segmentController struct {
+	sync.RWMutex
+	location     string
+	intervalRule IntervalRule
+	lst          []*segment
+}
+
+func newSegmentController(location string, intervalRule IntervalRule) *segmentController {
+	return &segmentController{
+		location:     location,
+		intervalRule: intervalRule,
+	}
+}
+
+func (sc *segmentController) get(segID uint16) *segment {
+	sc.RLock()
+	defer sc.RUnlock()
+	last := len(sc.lst) - 1
+	for i := range sc.lst {
+		s := sc.lst[last-i]
+		if s.id == segID {
+			return s
+		}
+	}
+	return nil
+}
+
+func (sc *segmentController) span(timeRange TimeRange) (ss []*segment) {
+	sc.RLock()
+	defer sc.RUnlock()
+	last := len(sc.lst) - 1
+	for i := range sc.lst {
+		s := sc.lst[last-i]
+		if s.Overlapping(timeRange) {
+			ss = append(ss, s)
+		}
+	}
+	return ss
+}
+
+func (sc *segmentController) segments() (ss []*segment) {
+	sc.RLock()
+	defer sc.RUnlock()
+	r := make([]*segment, len(sc.lst))
+	copy(r, sc.lst)
+	return r
+}
+
+func (sc *segmentController) Current() bucket.Reporter {
+	sc.RLock()
+	defer sc.RUnlock()
+	now := time.Now()
+	for _, s := range sc.lst {
+		if s.suffix == sc.intervalRule.Unit.Format(now) {
+			return s
+		}
+	}
+	// return the latest segment before now
+	if len(sc.lst) > 0 {
+		return sc.lst[len(sc.lst)-1]
+	}
+	return nil
+}
+
+func (sc *segmentController) Next() (bucket.Reporter, error) {
+	return sc.create(context.TODO(), sc.intervalRule.Unit.Format(
+		sc.intervalRule.NextTime(time.Now())))
+}
+
+func (sc *segmentController) open(ctx context.Context) error {
+	err := walkDir(
+		sc.location,
+		segPathPrefix,
+		func(suffix, absolutePath string) error {
+			_, err := sc.load(ctx, suffix, absolutePath)
+			return err
+		})
+	if err != nil {
+		return err
+	}
+	if sc.Current() == nil {
+		_, err = sc.create(ctx, sc.intervalRule.Unit.Format(time.Now()))
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (sc *segmentController) create(ctx context.Context, suffix string) (*segment, error) {
+	segPath, err := mkdir(segTemplate, sc.location, suffix)
+	if err != nil {
+		return nil, err
+	}
+	return sc.load(ctx, suffix, segPath)
+}
+
+func (sc *segmentController) load(ctx context.Context, suffix, path string) (seg *segment, err error) {
+	seg, err = openSegment(ctx, suffix, path, sc.intervalRule)
+	if err != nil {
+		return nil, err
+	}
+	{
```

Comment:
unnecessary pair of braces?

Reply:
it tends to reduce the scope of the lock